### ISSUE TITLE
Memory leak when using SequentialLR and ChainedLR schedulers

### ISSUE THREAD

**Issue by mlazos** [pytorch/pytorch#126131]
### 🐛 Describe the bug

Looks like memory is allocated in the caching allocator which doesn't get freed. I checked refcycle and it doesn't appear to be due to a cyclical reference in the python code. Repro below.

```py
import torch
from torch.optim.lr_scheduler import LambdaLR, ChainedScheduler, ConstantLR, SequentialLR
from torch.testing._internal.common_utils import CudaMemoryLeakCheck, TestCase, run_tests


LR_SCHEDULER_TO_KWARGS = {
    LambdaLR: {"lr_lambda": lambda x: 10},
    ChainedScheduler: {"schedulers": None},
    SequentialLR: {"schedulers": None, "milestones": [1, 2]},
    ConstantLR: {"factor": 0.001},
}


def create_scheduler(scheduler, optim):
    kwargs = LR_SCHEDULER_TO_KWARGS[scheduler]
    if "schedulers" in kwargs:
        kwargs["schedulers"] = [
            create_scheduler(ConstantLR, optim)
            for _ in range(2)
        ] + [create_scheduler(LambdaLR, optim)]

    if scheduler == ChainedScheduler:
        return scheduler(**kwargs)
    else:
        return scheduler(optim, **kwargs)

def chained_fn():
    from torch.testing._internal.common_utils import CudaMemoryLeakCheck
    with CudaMemoryLeakCheck(None,name="hi"):
        device="cuda:0"
        dtype=torch.float32
        optim_cls = torch.optim.ASGD
        kwargs = {'lr': 0.001, 'weight_decay': 0.1, 'maximize': True, 'capturable': True, 'foreach': False}
        scheduler_cls = ChainedScheduler
        print(scheduler_cls)
        print(kwargs)
        input = torch.ones([10, 10], device=device)
        model_eager = torch.nn.Sequential(
            *[torch.nn.Linear(10, 10, device=device) for _ in range(1)]
        )
        model_eager(input).sum().backward()
        opt_eager = optim_cls(model_eager.parameters(), **kwargs)
        scheduler_eager = create_scheduler(scheduler_cls, opt_eager)


def sequential_fn():
    from torch.testing._internal.common_utils import CudaMemoryLeakCheck
    with CudaMemoryLeakCheck(None,name="hi"):
        device="cuda:0"
        dtype=torch.float32
        optim_cls = torch.optim.ASGD
        kwargs = {'lr': 0.001, 'weight_decay': 0.1, 'maximize': True, 'capturable': True, 'foreach': False}
        scheduler_cls = SequentialLR
        print(scheduler_cls)
        print(kwargs)
        input = torch.ones([10, 10], device=device)
        model_eager = torch.nn.Sequential(
            *[torch.nn.Linear(10, 10, device=device) for _ in range(1)]
        )
        model_eager(input).sum().backward()
        opt_eager = optim_cls(model_eager.parameters(), **kwargs)
        scheduler_eager = create_scheduler(scheduler_cls, opt_eager)

# both sequential_fn and chained_fn fail
chained_fn()
```

### Versions

nightly:
a5c93a6899c657832944cd2eeb5069449e28dbea

**Comment by janeyx99** [comment]
I tried narrowing down the repro more as it's weird to me that this triggers even on construction without running any steps.

If I remove the nn module constructs and pass in rand tensors for params, we no longer get the hard error but just the warning:
```
def play_fun():
    device="cuda:0"
    dtype=torch.float32
    optim_cls = torch.optim.SGD
    kwargs = {}
    scheduler_cls = ChainedScheduler
    print(scheduler_cls)
    print(kwargs)
    # input = torch.ones([100, 100], device=device)
    # model_eager = torch.nn.Sequential(
    #     *[torch.nn.Linear(100, 100, device=device) for _ in range(1)]
    # )
    # model_eager(input).sum().backward()
    params = [torch.rand(512, 10, device=device) for _ in range(2)]
    opt_eager = optim_cls(params, **kwargs)
    scheduler_eager = create_scheduler(scheduler_cls, opt_eager)

def chained_fn():
    from torch.testing._internal.common_utils import CudaMemoryLeakCheck
    with CudaMemoryLeakCheck(None,name="chained"):
        play_fun()
```

Warning:
```
/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py:2197: UserWarning: CUDA caching allocator reports a memory leak not verified by the driver API in chained! Caching allocator allocated memory was 0 and is now reported as 40960 on device 0. CUDA driver allocated memory was 2960850944 and is now 2960850944.
  warnings.warn(msg)
```

BUT if we also delete the last scheduler construction line, we get no warning at all. I no longer know if this is high pri.

**Comment by janeyx99** [comment]
Removing the hi pri labels because the problem is in the repro (thanks @albanD for spotting)! If you add a deepcopy to the first line in create_scheduler:
```
kwargs = deepcopy(LR_SCHEDULER_TO_KWARGS[scheduler])
```
the memory "leak" goes away. So there was never a leak in the first place, but it seems that the test infra in compiled_optimizers.py has global state that holds onto these scheduler instances.

**Comment by mlazos** [comment]
I don't see why keeping the scheduler type as a key in a map should allocate any memory at all? These aren't actual instances, they're the type. The kwargs also don't allocate anything so I don't see why adding a deepcopy there fixes this. Can you further explain why that fixes this?

**Comment by mlazos** [comment]
> Removing the hi pri labels because the problem is in the repro (thanks @albanD for spotting)! If you add a deepcopy to the first line in create_scheduler:
> 
> ```
> kwargs = deepcopy(LR_SCHEDULER_TO_KWARGS[scheduler])
> ```
> 
> the memory "leak" goes away. So there was never a leak in the first place, but it seems that the test infra in compiled_optimizers.py has global state that holds onto these scheduler instances.

I don't see why keeping the scheduler type as a key in a map should allocate any memory at all? These aren't actual instances, they're the type. The kwargs also don't allocate anything so I don't see why adding a deepcopy there fixes this. Can you further explain why that fixes this?

**Comment by mlazos** [comment]
Update: I removed the global state entirely and still encounter the issue. [repro](https://gist.github.com/mlazos/b5cac7945f388cca071e95ff3963d8ea)

**Comment by janeyx99** [comment]
When I move the code under the context manager to a function and then call that function (similar to what i typed above), there is no failure. I think python may still be holding onto data otherwise?

**Comment by mlazos** [comment]
> When I move the code under the context manager to a function and then call that function (similar to what i typed above), there is no failure. I think python may still be holding onto data otherwise?

Okay let me try this, what's extra strange is that in the actual test cases the mem leak checker is outside of the test case function, so I would think this would apply there as well. I'll keep playing with it to see.

**Comment by masnesral** [comment]
2025Q1 issue scrubbing: @mlazos I can't pretend I understand, but looks like you were going to try something per your last update? Anything to do here? Leave this open for now?
